### PR TITLE
Kafka Connect Sink - to send messages from Kafka to IoT devices

### PR DESCRIPTION
**Pull request by varunpuranik**
Added support for a Kafka Connect Azure IoT Hub Sink that sends messages from Kafka topics to IoT devices. The messages need to be in a particular format as specified in the documentation. This does not yet include sending feedback from the sent messages back to some Kafka topic.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,190 +1,25 @@
...
 
-Kafka Connect Azure IoT Hub is a Kafka source connector for  pumping data from
-[Azure IoT Hub](https://azure.microsoft.com/en-us/services/iot-hub/) to [Apache Kafka](https://kafka.apache.org/). This
-allows getting the telemetry data sent by Azure IoT Hub connected devices to your Kafka installation, so that it can
- then be consumed by Kafka consumers down the stream.
+Kafka Connect Azure IoT Hub consists of 2 connectors - a source connector and a sink connector. The source connector
+is used to pump data from [Azure IoT Hub](https://azure.microsoft.com/en-us/services/iot-hub/) to
+[Apache Kafka](https://kafka.apache.org/), whereas the sink connector reads messages from Kafka and sends them to IoT
+ devices via [Azure IoT Hub](https://azure.microsoft.com/en-us/services/iot-hub/). When used in tandem, the 2
+ connectors allow communicating with IoT devices by simply posting and reading messages to Kafka topics. This should
```
Comment (dluc) [inline README.md]: possible typo: "posting and reading messages to/from ..."

File: README.md [hunk 2]
```diff
@@ -1,190 +1,25 @@
...
+ makie it easier for open source systems and other systems that already interface with Kafka to communicate with
```
Comment (dluc) [inline README.md]: typo "makie"

File: README_Sink.md [hunk 3]
```diff
@@ -0,0 +1,182 @@
+# Kafka Connect Sink Connector for Azure IoT Hub
+________________________
+
+The Sink connector allows you to send messages to Azure IoT devices by simply posting messages to Kafka topics. The
+connector sends the messages to Azure IoT Hub, which in turn forwards them to the right devices. The
+messages need to be in a specific format (details below), that allow the connector to extract the information
```
Comment (dluc) [inline README_Sink.md]: typo "allows"

File: README_Sink.md [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+connector.class=com.microsoft.azure.iot.kafka.connect.sink.IotHubSinkConnector
+name=AzureIotHubSinkConnector
+tasks.max=1
+topics=testtopic
+IotHub.ConnectionString=HostName=Test.azure-devices.net;SharedAccessKeyName=service;SharedAccessKey=aBCdeBb5HffTTs/J9ikdcqab1JNMB0ot=
+IotHub.MessageDeliveryAcknowledgement=None
+```
+
+> Note: At the moment, the Sink connector does not support getting feedback from the devices on the status of the
+> messages (whether accepted or rejected). If you want to get feedback, you will need to do it manually. Please refer
```
Comment (dluc) [inline README_Sink.md]: accepted/rejected/expired

File: README_Sink.md [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+> to the documentation [here](https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-java-java-c2d) on how to do that.
+
+
+### Data format
+
+The sink connector expects the messages in the Kafka topic to have the schema below, so that it can extract the
+information required to send the message to the IoT device. If a record is not in the expected format, the connector
+will throw an exception.
```
Comment (dluc) [inline README_Sink.md]: perhaps we should clarify if the connector stops or continues to work, ignoring the poison message

File: README_Source.md [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+
+Here is a sample connect-iothub-source.properties file. (Some of these values are not configurable, and hence omitted
+  from the list above)
+```
+connector.class=com.microsoft.azure.iot.kafka.connect.IotHubSourceConnector
+name=AzureIotHubConnector
+tasks.max=1
+Kafka.Topic=IotTopic
+IotHub.EventHubCompatibleName=iothub-toketi
+IotHub.EventHubCompatibleEndpoint=sb://iothub-ns-toketi-001.servicebus.windows.net/
```
Comment (dluc) [inline README_Source.md]: here I would use a more generic name, e.g. sb://my-iothub-001.servicebus.windows.net/

File: README_Source.md [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+IotHub.AccessKeyName=service
+IotHub.AccessKeyValue=6XdRSFB9H61f+N3uOdBJiKwzeqbZUj1K//T2jFyewN4=
```
Comment (dluc) [inline README_Source.md]: please make sure that this key and the others in the PR, are not valid

File: build.sbt [hunk 8]
```diff
@@ -19,6 +19,7 @@ libraryDependencies ++= {
   val scalaTestVersion = "3.0.0"
   val configVersion = "1.3.1"
   val json4sVersion = "3.5.0"
+  val iotHubServiceClientVersion = "1.0.11"
```
Comment (dluc) [inline build.sbt]: the latest is 1.0.12, with a new namespace+name

"com.microsoft.azure.sdk.iot" % "iot-service-client" % "1.0.12"

File: src/main/scala/com/microsoft/azure/iot/kafka/connect/sink/C2DMessage.scala [hunk 9]
```diff
@@ -0,0 +1,9 @@
+/*
+ * // Copyright (c) Microsoft. All rights reserved.
+ */
+
+package com.microsoft.azure.iot.kafka.connect.sink
+
+import java.util.Date
+
+case class C2DMessage(messageId: String, message: String, deviceId: String, expiryTime: Option[Date] = None)
```
Comment (dluc) [inline src/main/scala/com/microsoft/azure/iot/kafka/connect/sink/C2DMessage.scala:9]: we could support the ACK modes, see com.microsoft.azure.sdk.iot.service.sdk.DeliveryAcknowledgement: None, NegativeOnly, PositiveOnly, Full

File: src/main/scala/com/microsoft/azure/iot/kafka/connect/sink/C2DMessageConverter.scala [hunk 10]
```diff
@@ -0,0 +1,110 @@
+/*
```
Comment (dluc) [inline src/main/scala/com/microsoft/azure/iot/kafka/connect/sink/C2DMessageConverter.scala:1]: nitpick: please keep public methods before the private ones

File: src/main/scala/com/microsoft/azure/iot/kafka/connect/sink/IotHubMessageSender.scala [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+
+  private val serviceClient = ServiceClient.createFromConnectionString(connectionString,
+    IotHubServiceClientProtocol.AMQPS)
+  this.serviceClient.open()
+
+  def sendMessage(deviceId: String, message: Message): CompletableFuture[Void] = {
+    this.serviceClient.sendAsync(deviceId, message)
+  }
+
+  def close(): Unit = {
```
Comment (dluc) [inline src/main/scala/com/microsoft/azure/iot/kafka/connect/sink/IotHubMessageSender.scala:19]: could this return a Future too, like the previous method?

File: src/main/scala/com/microsoft/azure/iot/kafka/connect/sink/IotHubSinkConnector.scala [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+
+class IotHubSinkConnector extends SinkConnector with LazyLogging with JsonSerialization {
+
+  private[this] var props: Map[String, String] = _
+
+  override def taskClass(): Class[_ <: Task] = classOf[IotHubSinkTask]
+
+  override def taskConfigs(maxTasks: Int): util.List[util.Map[String, String]] = {
+    val configList = (1 to maxTasks).map(c => this.props.asJava).toList.asJava
+    configList
```
Comment (dluc) [inline src/main/scala/com/microsoft/azure/iot/kafka/connect/sink/IotHubSinkConnector.scala]: nitpick: you could avoid the local val and use a single line of code

**Review by dluc — APPROVED** [review]
Some minor changes it would be nice to have before releasing

File: src/main/scala/com/microsoft/azure/iot/kafka/connect/sink/C2DMessage.scala (hunk 9 above)
Comment (varunpuranik) [reply src/main/scala/com/microsoft/azure/iot/kafka/connect/sink/C2DMessage.scala:9]: These are supported as config values.

File: src/main/scala/com/microsoft/azure/iot/kafka/connect/sink/IotHubMessageSender.scala (hunk 11 above)
Comment (varunpuranik) [reply src/main/scala/com/microsoft/azure/iot/kafka/connect/sink/IotHubMessageSender.scala:19]: Changes this to use the sync version, as the async versions has some bugs.

File: README_Source.md (hunk 7 above)
Comment (varunpuranik) [reply README_Source.md]: Yes, they are all dummy.

**Pull request merged**